### PR TITLE
CIFSContext pooling

### LINKED ISSUE #2086: SMB copy operation sometimes lead to failure
**Describe the bug**
For few files or maybe under certain scenario, SMB copy fails with following log

**Additional context**
Add any other context about the problem here
```
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err: jcifs.smb.SmbException: No more connections can be made to this remote computer at this time because there are already as many connections as the computer can accept.
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at jcifs.smb.SmbTransportImpl.checkStatus(SmbTransportImpl.java:1391)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at jcifs.smb.SmbTransportImpl.checkStatus(SmbTransportImpl.java:1573)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at jcifs.smb.SmbTransportImpl.sendrecv(SmbTransportImpl.java:1023)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at jcifs.smb.SmbTransportImpl.send(SmbTransportImpl.java:1539)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at jcifs.smb.SmbTransportImpl.send(SmbTransportImpl.java:1503)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at jcifs.smb.SmbSessionImpl.sessionSetupSMB1(SmbSessionImpl.java:1026)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at jcifs.smb.SmbSessionImpl.sessionSetup(SmbSessionImpl.java:486)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at jcifs.smb.SmbSessionImpl.send(SmbSessionImpl.java:369)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at jcifs.smb.SmbSessionImpl.send(SmbSessionImpl.java:347)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at jcifs.smb.SmbTreeImpl.treeConnect(SmbTreeImpl.java:611)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at jcifs.smb.SmbTreeConnection.connectTree(SmbTreeConnection.java:614)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at jcifs.smb.SmbTreeConnection.connectHost(SmbTreeConnection.java:568)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at jcifs.smb.SmbTreeConnection.connectHost(SmbTreeConnection.java:489)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at jcifs.smb.SmbTreeConnection.connect(SmbTreeConnection.java:465)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at jcifs.smb.SmbTreeConnection.connectWrapException(SmbTreeConnection.java:426)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at jcifs.smb.SmbFile.ensureTreeConnected(SmbFile.java:551)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at jcifs.smb.SmbFileInputStream.<init>(SmbFileInputStream.java:96)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at jcifs.smb.SmbFileInputStream.<init>(SmbFileInputStream.java:86)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at jcifs.smb.SmbFile.getInputStream(SmbFile.java:2104)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.filesystem.HybridFile.getInputStream(HybridFile.java:943)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.filesystem.files.GenericCopyUtil.startCopy(GenericCopyUtil.java:120)
2020-11-13 12:59:34.120 31952-32151/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.filesystem.files.GenericCopyUtil.copy(GenericCopyUtil.java:363)
2020-11-13 12:59:34.121 31952-32151/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.services.CopyService$DoInBackground$Copy.copyFiles(CopyService.java:513)
2020-11-13 12:59:34.121 31952-32151/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.services.CopyService$DoInBackground$Copy.execute(CopyService.java:402)
2020-11-13 12:59:34.121 31952-32151/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.services.CopyService$DoInBackground.doInBackground(CopyService.java:266)
2020-11-13 12:59:34.121 31952-32151/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.services.CopyService$DoInBackground.doInBackground(CopyService.java:231)
2020-11-13 12:59:34.121 31952-32151/com.amaze.filemanager.debug W/System.err:     at android.os.AsyncTask$3.call(AsyncTask.java:378)
2020-11-13 12:59:34.121 31952-32151/com.amaze.filemanager.debug W/System.err:     at java.util.concurrent.FutureTask.run(FutureTask.java:266)
2020-11-13 12:59:34.121 31952-32151/com.amaze.filemanager.debug W/System.err:     at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:289)
2020-11-13 12:59:34.121 31952-32151/com.amaze.filemanager.debug W/System.err:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
2020-11-13 12:59:34.121 31952-32151/com.amaze.filemanager.debug W/System.err:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
2020-11-13 12:59:34.121 31952-32151/com.amaze.filemanager.debug W/System.err:     at java.lang.Thread.run(Thread.java:919)
2020-11-13 12:59:34.127 31952-32151/com.amaze.filemanager.debug W/System.err: java.io.IOException: Stream closed
2020-11-13 12:59:34.127 31952-32151/com.amaze.filemanager.debug W/System.err:     at java.io.BufferedInputStream.getInIfOpen(BufferedInputStream.java:161)
2020-11-13 12:59:34.127 31952-32151/com.amaze.filemanager.debug W/System.err:     at java.io.BufferedInputStream.fill(BufferedInputStream.java:248)
2020-11-13 12:59:34.127 31952-32151/com.amaze.filemanager.debug W/System.err:     at java.io.BufferedInputStream.read1(BufferedInputStream.java:288)
2020-11-13 12:59:34.127 31952-32151/com.amaze.filemanager.debug W/System.err:     at java.io.BufferedInputStream.read(BufferedInputStream.java:347)
2020-11-13 12:59:34.127 31952-32151/com.amaze.filemanager.debug W/System.err:     at java.nio.channels.Channels$ReadableByteChannelImpl.read(Channels.java:386)
2020-11-13 12:59:34.127 31952-32151/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.filesystem.files.GenericCopyUtil.doCopy(GenericCopyUtil.java:433)
2020-11-13 12:59:34.127 31952-32151/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.filesystem.files.GenericCopyUtil.copyFile(GenericCopyUtil.java:377)
2020-11-13 12:59:34.127 31952-32151/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.filesystem.files.GenericCopyUtil.startCopy(GenericCopyUtil.java:312)
2020-11-13 12:59:34.127 31952-32151/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.filesystem.files.GenericCopyUtil.copy(GenericCopyUtil.java:363)
2020-11-13 12:59:34.127 31952-32151/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.services.CopyService$DoInBackground$Copy.copyFiles(CopyService.java:513)
2020-11-13 12:59:34.127 31952-32151/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.services.CopyService$DoInBackground$Copy.execute(CopyService.java:402)
2020-11-13 12:59:34.127 31952-32151/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.services.CopyService$DoInBackground.doInBackground(CopyService.java:266)
2020-11-13 12:59:34.127 31952-32151/com.amaze.filemanager.debug W/System.err:     at com.amaze.filemanager.asynchronous.services.CopyService$DoInBackground.doInBackground(CopyService.java:231)
2020-11-13 12:59:34.127 31952-32151/com.amaze.filemanager.debug W/System.err:     at android.os.AsyncTask$3.call(AsyncTask.java:378)
2020-11-13 12:59:34.127 31952-32151/com.amaze.filemanager.debug W/System.err:     at java.util.concurrent.FutureTask.run(FutureTask.java:266)
2020-11-13 12:59:34.127 31952-32151/com.amaze.filemanager.debug W/System.err:     at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:289)
2020-11-13 12:59:34.127 31952-32151/com.amaze.filemanager.debug W/System.err:     at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
2020-11-13 12:59:34.128 31952-32151/com.amaze.filemanager.debug W/System.err:     at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
2020-11-13 12:59:34.128 31952-32151/com.amaze.filemanager.debug W/System.err:     at java.lang.Thread.run(Thread.java:919)
2020-11-13 12:59:34.128 31952-32151/com.amaze.filemanager.debug D/GenericCopyUtil: I/O Error!

```

### PR DESCRIPTION
**Pull request by TranceLove**
#### Issue tracker   
Fixes will automatically close the related issue

Fixes #
-or-   
Addresses #2086

#### Release  
Addresses release/3.6
  
#### Test cases
- [x] Covered
  
#### Manual testing
- [x] Done  
  
If yes,  
- Device: Pixel 3 emulator
- OS: Android 10

#### Build tasks success  
Successfully running following tasks on local 
- [x] `./gradlew assembledebug`
- [x] `./gradlew spotlessCheck`

#### Related PR  
Related to PR #

#### Additional Info
Restrict redundant creation of CIFSContext/BaseContext instances, to hopefully fix #2086.

BaseContexts are saved using in a map, similar to SshConnectionPool, using `smb://<credentials>@<host><additional parameters as query string>` as key, to allow reusing connections if same credentials (and settings) are used on connecting SMB server.

### CODE REVIEW

**Comment by VishalNehra** [comment]
Changes seem fine. Will test once.

File: app/src/main/java/com/amaze/filemanager/filesystem/smb/CifsContexts.kt [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            val context = factory.create(extraProperties)
+            contexts[basePathKey] = context
+            context
+        }
+    }
+
+    /**
+     * Interface for factory classes creating BaseContext instances.
+     */
+    interface Factory {
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/filesystem/smb/CifsContexts.kt:101]: Factory is redundant, we should probably add it when we need it.

**Review by EmmanuelMess — CHANGES_REQUESTED** [review]

**Comment by VishalNehra** [comment]
Awesome work @TranceLove operations have become much faster with these changes!

**Review by VishalNehra — APPROVED** [review]

**Review by EmmanuelMess — APPROVED** [review]
Had to apply #2180.

**Pull request merged**